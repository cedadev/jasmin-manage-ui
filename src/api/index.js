import { useEffect, useState } from 'react';

import {
    apiFetch,
    createContextForEndpoint,
    createContextForResource,
    useResource,
    useInstance,
    useFetchPoint
} from '../rest-resource';


// Create a context for the current user and export the hook
const CurrentUser = createContextForEndpoint("/api/me/");
const CurrentUserProvider = CurrentUser.Provider;
export const useCurrentUser = CurrentUser.hook;


// All other resources require authentication before they can be fetched
// This higher-order component wraps a provider so that it is only mounted once
// the user is successfully initialised
const withAuthentication = Provider => ({ children }) => {
    const currentUser = useCurrentUser();
    // Only wrap the children in the provider if the user is authenticated
    return currentUser.initialised ? <Provider>{children}</Provider> : children;
};


// Create a context each for the top-level, read-only resources that don't change
// often and we want to share across the component tree
const Categories = createContextForResource("/api/categories/");
const Consortia = createContextForResource("/api/consortia/");
const Resources = createContextForResource("/api/resources/");

// Wrap the providers with the higher-order component for authentication
const CategoriesProvider = withAuthentication(Categories.Provider);
const ConsortiaProvider = withAuthentication(Consortia.Provider);
const ResourcesProvider = withAuthentication(Resources.Provider);

// Export the hooks for each resource
export const useCategories = Categories.hook;
export const useConsortia = Consortia.hook;
export const useResources = Resources.hook;

// Export a single provider that provides all the top-level resources
export const Provider = ({ children }) => (
    <CurrentUserProvider>
        <CategoriesProvider>
            <ConsortiaProvider>
                <ResourcesProvider>
                    {children}
                </ResourcesProvider>
            </ConsortiaProvider>
        </CategoriesProvider>
    </CurrentUserProvider>
);


// Export a hook for loading a consortium by id
export const useConsortium = (id, options) => useInstance(`/api/consortia/${id}/`, options);

// Export hooks for accessing projects
// We don't treat the project list as a top-level shared resource because:
//   1. The list, or objects within it, changes relatively often
//   2. Projects can be loaded by id that are not in the list, e.g. when the user
//      is a consortium manager, which reduces the value in caching the list
export const useProjects = options => useResource("/api/projects/", options);
export const useProject = (id, options) => useInstance(`/api/projects/${id}/`, options);

// Export hook for loading requirement by id
export const useRequirement = (id, options) => useInstance(`/api/requirements/${id}/`, options);

// Export hook for loading service by id
export const useService = (id, options) => useInstance(`/api/services/${id}/`, options);

// Export a function for joining a project with an invite code
export const joinProject = async code => apiFetch('/api/join/', { method: 'POST', data: { code } });

// Export hook for loading category by id
export const useCategory = (id, options) => useInstance(`/api/categories/${id}/`, options);

// Export hook for loading resource by id
export const useResourceType = (id, options) => useInstance(`/api/resources/${id}/`, options);


// Export a hook for using a project event stream
export const useProjectEvents = (project, options) => {
    const { fetchPoint = true, autoFetch = true } = options || {};

    const [initialised, setInitialised] = useState(false);
    const [dirty, setDirty] = useState(autoFetch);
    const [fetching, setFetching] = useState(false);
    const [fetchError, setFetchError] = useState(null);
    const [data, setData] = useState([]);

    // Get the events URL from the project links
    const eventsUrl = project.data._links.events;

    // Build the stream object with fetchable methods
    const stream = {
        initialised,
        dirty,
        fetching,
        fetchError,
        data,
        // This function loads more events
        fetch: () => {
            // Build the URL to use
            const url = new URL(eventsUrl);
            // If there is a most recent event, use its time for the since parameter
            const event = data.find(_ => true);
            if( event ) url.searchParams.append("since", event.created_at);
            // Mark the fetch as in progress
            setFetching(true);
            // In order to prevent state updates after a component has been unmounted,
            // we need to be able to cancel fetches
            // To do this, we can use an abort controller
            const controller = new AbortController();
            // Start the fetch using the signal from the controller
            apiFetch(url, { signal: controller.signal })
                // If the promise resolves, update the state with the new data
                .then(newData => {
                    // Prepend the data to the existing data
                    setData([...newData, ...data]);
                    setInitialised(true);
                    setFetching(false);
                    setDirty(false);
                    setFetchError(null);
                })
                // If the promise rejects, update the state with the error
                .catch(error => {
                    // If the error is an abort error, it is because the component that
                    // owns the data has been unmounted so don't update the state
                    if( error.name === "AbortError" ) return;
                    setFetching(false);
                    setDirty(false);
                    setFetchError(error);
                });
            // Return the abort function to allow the fetch to be cancelled from outside
            return () => controller.abort();
        },
        // This function marks the stream as dirty, causing more data to be fetched
        markDirty: () => setDirty(true),
        // This function resets the stream, clearing all the data and triggering a re-fetch
        reset: () => {
            setInitialised(false);
            setDirty(true);
            setFetching(false);
            setFetchError(null);
            setData([]);
        }
    };

    // Hooks cannot be conditional, so we always have to call the fetch point hook
    // However we can effectively disable it by making sure it never sees a dirty stream
    useFetchPoint({ ...stream, dirty: fetchPoint && stream.dirty });

    // We need to be a fetch point for the event stream
    return stream;
};
