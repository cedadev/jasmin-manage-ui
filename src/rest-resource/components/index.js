// Export the delete button and status components directly
export * from './InstanceDeleteButton';
export * from './Status';

// Export the form pieces in their own namespace
import * as Form from './ResourceForm';
export { Form };
