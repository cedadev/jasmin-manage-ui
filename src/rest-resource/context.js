import React, { createContext, useContext, useEffect } from 'react';

import { useEndpoint, useResource, useInstance } from './hooks';


// This higher order function turns a hook into a context, allowing the hook
// state to be shared between components
const createContextForHook = hookFunction => (url, options = {}) => {
    // Make a new context
    const Context = createContext();
    // Return a provider and hook that use the context to share the hook state
    return {
        // The provider is an anchor for the data and also a fetch point
        // However, we do not auto-fetch the data as soon as the provider as mounted
        Provider: ({ children }) => {
            const fetchable = hookFunction(url, { ...options, autoFetch: false });
            return <Context.Provider value={fetchable}>{children}</Context.Provider>;
        },
        // The first time a component using the hook is mounted, the data for the fetchable
        // is initialised if it has not already been initialised
        hook: (autoFetch = true) => {
            const fetchable = useContext(Context);
            useEffect(
                // If the fetchable is not initialised, trigger a fetch by marking it as dirty
                // This should not trigger multiple fetches
                () => { if( autoFetch && !fetchable.initialised ) fetchable.markDirty(); },
                []
            );
            return fetchable;
        }
    };
};


export const createContextForEndpoint = createContextForHook(useEndpoint);
export const createContextForResource = createContextForHook(useResource);
export const createContextForInstance = createContextForHook(useInstance);
