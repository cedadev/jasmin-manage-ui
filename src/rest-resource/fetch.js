import Cookies from 'js-cookie';


class HttpError extends Error {
    constructor(status, statusText, responseText) {
        super(statusText);
        this.name = this.constructor.name;
        this.status = status;
        this.statusText = statusText;
        this.responseText = responseText;
    }

    json() {
        return JSON.parse(this.responseText)
    }
}


// Utility function that asynchronously waits for the specified duration
const sleep = (ms) => (new Promise(resolve => setTimeout(resolve, ms)));


const apiFetch = async (url, {
    method = 'GET',
    headers = {},
    data,
    body,
    ...options
} = {}) => {
    // Populate the required headers for the request
    const defaultHeaders = {};
    // For POST/PUT/DELETE, declare the content type and include the CSRF token if present
    if( ['POST', 'PUT', 'DELETE'].includes(method.toUpperCase()) ) {
        defaultHeaders['Content-Type'] = 'application/json';
        const csrfToken = Cookies.get('csrftoken');
        if( csrfToken ) defaultHeaders['X-CSRFToken'] = csrfToken;
    }
    // Make the actual request, injecting the cookie credentials and headers
    const response = await fetch(url, {
        method,
        headers: { ...defaultHeaders, ...headers },
        // If data is given, encode it as JSON and use it as the request body
        // If not, use the given body
        body: data ? JSON.stringify(data) : body,
        // Always include credentials
        credentials: 'include',
        // Forward any other options that were supplied
        ...options
    });
    // For a 204, just return
    if( response.status === 204 ) return;
    // Any other successful response should be JSON
    if( response.ok ) return await response.json();
    // An error response may not be JSON, so read the response as text
    const responseText = await response.text();
    throw new HttpError(response.status, response.statusText, responseText);
};


// Export method-specific versions of apiFetch
export default apiFetch;
