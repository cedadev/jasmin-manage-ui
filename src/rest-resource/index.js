export * from './components';
export * from './context';
export * from './hooks';

import apiFetch from './fetch';
export { apiFetch };
