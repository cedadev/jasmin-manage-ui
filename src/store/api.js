import Cookies from 'js-cookie';

import { withScope } from './utils';


class HttpError extends Error {
    constructor(status, statusText, responseText) {
        super(statusText);
        this.name = this.constructor.name;
        this.status = status;
        this.statusText = statusText;
        this.responseText = responseText;
    }

    json() {
        return JSON.parse(this.responseText)
    }
}


class InvalidOperationError extends Error {
    constructor(operationName) {
        super(`Invalid operation: ${operationName}`);
        this.name = this.constructor.name;
    }
}


// Utility function that asynchronously waits for the specified duration
const sleep = (ms) => (new Promise(resolve => setTimeout(resolve, ms)));


const apiFetch = async (url, options = {}) => {
    // Populate the required headers for the request
    const headers = {};
    // For POST/PUT/DELETE, declare the content type and include the CSRF token if present
    const method = options.method || 'GET';
    if( ['POST', 'PUT', 'DELETE'].includes(method.toUpperCase()) ) {
        headers['Content-Type'] = 'application/json';
        const csrfToken = Cookies.get('csrftoken');
        if( csrfToken ) headers['X-CSRFToken'] = csrfToken;
    }
    // Make the actual request, injecting the cookie credentials and headers
    const response = await fetch(url, { ...options, headers, credentials: 'include' });
    // For a 204, just return
    if( response.status === 204 ) return;
    // Any other successful response should be JSON
    if( response.ok ) return await response.json();
    // An error response may not be JSON, so read the response as text
    const responseText = await response.text();
    // If the response indicates that the request requires authentication, redirect to auth
    if( [401, 403].includes(response.status) ) {
        // These responses will be JSON and should have a code
        const code = JSON.parse(responseText).code;
        // Some auth schemes use 401 to indicate lack of credentials, and some use
        // 403 with a code of not_authenticated - we need to catch both
        if( response.status === 401 || code === "not_authenticated" ) {
            // Get the path to redirect back to from the current location
            const currentPath = new URL(window.location.href).pathname;
            window.location = `/auth/login/?next=${currentPath}`;
            // Wait for the browser to redirect
            await sleep(2000);
        }
    }
    throw new HttpError(response.status, response.statusText, responseText);
};


const apiGet = (url, options) => apiFetch(url, options);
const apiPost = (url, data, options = {}) => apiFetch(url, {
    ...options,
    method: 'POST',
    body: data ? JSON.stringify(data) : undefined
});
const apiPut = (url, data, options = {}) => apiFetch(url, {
    ...options,
    method: 'PUT',
    body: data ? JSON.stringify(data) : undefined
});
const apiDelete = (url, options = {}) => apiFetch(url, {
    ...options,
    method: 'DELETE'
});


export const newEndpoint = path => (mutate, getState) => ({
    // This is true once the endpoint has been fetched once
    initialised: false,
    // Indicates whether a fetch is currently in progress
    fetching: false,
    // The error from the last fetch, if there is one
    fetchError: null,
    initialise: async () => {
        const resource = getState();
        // If the resource is not initialised, trigger a fetch
        if( !resource.initialised ) await resource.fetch();
    },
    fetch: async () => {
        // If there is a fetch in progress, wait for the fetch to finish in a non-blocking way
        if( getState().fetching ) {
            while( getState().fetching ) await sleep(1);
            return;
        }
        // Otherwise, initiate a fetch
        mutate(state => { state.fetching = true; });
        try {
            // Fetch the resources from the API
            const result = await apiGet(path);
            // Update the state
            mutate(state => {
                // We are no longer fetching
                state.fetching = false;
                // We are also initialised
                state.initialised = true;
                // We also did a fetch without errors
                state.fetchError = null;
                // Just store the data as-is
                state.data = result;
            });
        }
        catch(error) {
            // For an error response, mark the fetch as over and store the error
            mutate(state => {
                state.fetching = false;
                state.fetchError = error;
            });
            // Rethrow the error
            throw error;
        }
    }
});


const defaultResource = {
    // This is true once there has been at least one successful fetch
    initialised: false,
    // Indicates whether a fetch is currently in progress
    fetching: false,
    // The error from the last fetch, if there is one
    fetchError: null,
    // Indicates whether a create is currently in progress
    creating: false,
    // The current data as a map of id => instance
    data: {},
    // By default, initialise is a no-op
    initialise: async () => {/* NOOP */},
    // By default, all other operations are not valid
    fetch: async () => { throw new InvalidOperationError('fetch'); },
    create: async () => { throw new InvalidOperationError('create'); },
    aggregate: () => { throw new InvalidOperationError('aggregate'); }
};


export const newResource = resourcePath => (mutate, getState, getParent) => ({
    ...defaultResource,
    // If this is a nested resource, this will return the parent resource
    parent: () => (getParent ? getParent() : undefined),
    // Initialises the resource
    initialise: async () => {
        const resource = getState();
        // If the resource is not initialised, trigger a fetch
        if( !resource.initialised ) await resource.fetch();
    },
    // Fetches all the instances for the resource
    fetch: async () => {
        // If there is a fetch in progress, wait for the fetch to finish in a non-blocking way
        if( getState().fetching ) {
            while( getState().fetching ) await sleep(1);
            return;
        }
        // Otherwise, initiate a fetch
        mutate(state => { state.fetching = true; });
        try {
            // Fetch the resources from the API
            const resourceList = await apiGet(resourcePath);
            // Update the state
            mutate(state => {
                // We are no longer fetching
                state.fetching = false;
                // We are also initialised
                state.initialised = true;
                // We also did a fetch without errors
                state.fetchError = null;
                // The data is stored as a dictionary of instances
                // Create or update the instance for each item in the returned list
                resourceList.forEach(resource => {
                    state.data[resource.id] = state.data.hasOwnProperty(resource.id) ?
                        // If there is an existing instance, update it with the new data
                        Object.assign(state.data[resource.id].data, resource) :
                        // If there is no existing instance, make a new one
                        newInstance(resource)(...withScope(mutate, getState, ["data", resource.id]));
                });
                // Remove items that are no longer in the list
                //   (remember that Javascript objects implicitly convert keys to strings)
                const nextIds = resourceList.map(resource => resource.id.toString());
                Object.keys(state.data).filter(id => !nextIds.includes(id)).forEach(id => { delete state.data[id]; });
            });
        }
        catch(error) {
            // For an error response, mark the fetch as over and store the error
            mutate(state => {
                state.fetching = false;
                state.fetchError = error;
            });
            // Rethrow the error
            throw error;
        }
    },
    // Creates a new instance of the resource
    create: async data => {
        // Mark the create as in progress
        mutate(state => { state.creating = true; });
        try {
            // Attempt to create the resource
            const resource = await apiPost(resourcePath, data);
            // Update the state if the create was successful
            mutate(state => {
                // We are no longer creating
                state.creating = false;
                // Add the new resource to the data
                state.data[resource.id] = newInstance(resource)(...withScope(mutate, getState, ["data", resource.id]));
            });
        }
        catch(error) {
            // Mark the create as over before re-throwing the error
            mutate(state => { state.creating = false; });
            throw error;
        }
    },
    // Initialise nested resources to an arbitrary depth using the URLs from links
    initNested: async (...keys) => {
        // Before the nested resources can be initialised, we need to be initialised
        await getState().initialise();
        // If no keys are given, there is nothing more to do
        if( keys.length === 0 ) return;
        // Otherwise, initialise the nested resources for each instance that was loaded
        const [firstKey, ...otherKeys] = keys;
        await Promise.all(
            Object.values(getState().data).map(instance =>
                instance.nested(firstKey).initNested(...otherKeys)
            )
        );
    },
    // Returns an aggregated resource across several levels of nesting
    aggregate: (...keys) => {
        const resource = getState();
        // If there are no keys, the aggregate resource is this resource
        if( keys.length === 0 ) return resource;
        // If this resource is not initialised, then there is nothing to aggregate
        // But the aggregate resource should also be considered not initialised
        if( !resource.initialised ) {
            return {
                ...defaultResource,
                // Initialising the aggregate means initialising the nested resources
                initialise: () => getState().initNested(...keys)
            };
        }
        // For each item in our data, use the first key to get a nested resource then
        // aggregate the remaining keys over the nested resource
        // Then collapse those resources to get the aggregate state
        const [firstKey, ...otherKeys] = keys;
        const aggregateState = Object.values(resource.data)
            .map(instance => instance.nested(firstKey).aggregate(...otherKeys))
            .reduce(
                (state1, state2) => ({
                    ...state1,
                    initialised: state1.initialised && state2.initialised,
                    fetching: state1.fetching || state2.fetching,
                    fetchError: state1.fetchError || state2.fetchError,
                    data: { ...state1.data, ...state2.data }
                }),
                // If there are no aggregates, we are initialised
                { ...defaultResource, initialised: true }
            );
        // Return a resource - aggregate resources only support initialise
        return {
            ...aggregateState,
            // Initialising the aggregate means initialising the nested resources
            initialise: () => getState().initNested(...keys)
        };
    }
});


export const newInstance = (initialData) => (mutate, getState, getParent) => ({
    // Indicates if the instance is currently being fetched
    fetching: false,
    // Indicates if the instance is currently being updated
    updating: false,
    // Indicates if the instance is currently being deleted
    deleting: false,
    // The current data for the instance
    data: initialData,
    // Initialise a nested resource for each key in the _links (that isn't self)
    // even if they aren't actually nested resources
    // Lazily loading these on demand is a major pain
    nestedResources: Object.assign(
        {},
        ...Object.entries(initialData._links)
            .filter(([name, _]) => !["self", "_links"].includes(name))
            .map(([name, url]) => ({
                [name]: newResource(url)(
                    // The resource requires scoped mutate and getState
                    ...withScope(mutate, getState, ["nestedResources", name])
                )
            }))
    ),
    // Returns the resource that the instance belongs to
    resource: getParent,
    // Returns a named nested resource
    nested: name => getState().nestedResources[name],
    // Update the instance with new data
    update: async data => {
        // Mark the update as in progress
        mutate(state => { state.updating = true; });
        try {
            const selfUrl = getState().data._links.self;
            // Attempt to update the resource
            const resource = await apiPut(selfUrl, data);
            // Update the state if the update was successful
            mutate(state => {
                state.updating = false;
                Object.assign(state.data, resource);
            });
        }
        catch(error) {
            // Mark the update as over before re-throwing the error
            mutate(state => { state.updating = false; });
            throw error;
        }
    },
    // Delete the instance
    delete: async () => {
        // Mark the delete as in progress
        mutate(state => { state.deleting = true; });
        try {
            const { id, _links: { self: selfUrl } } = getState().data;
            // Attempt to delete the resource
            await apiDelete(selfUrl);
            // Update the state if the delete was successful
            // We have to remove the instance from the parent resource
            mutate((state, parent) => {
                state.deleting = false;
                delete parent.data[id];
            });
        }
        catch(error) {
            // Mark the delete as over before re-throwing the error
            mutate(state => { state.deleting = false; });
            throw error;
        }
    },
    // Executes a named action on the resource
    executeAction: async (name, data) => {
        // Mark the resource as updating
        mutate(state => { state.updating = true; });
        try {
            // Get the URL for the action
            const actionUrl = getState().data._links[name];
            // Attempt to create the resource
            const resource = await apiPost(actionUrl, data);
            // Update the state if the action was successful
            mutate(state => {
                // We are no longer updating
                state.updating = false;
                // Update the resource data
                Object.assign(state.data, resource);
            });
        }
        catch(error) {
            // Mark the update as over before re-throwing the error
            mutate(state => { state.updating = false; });
            throw error;
        }
    }
});
