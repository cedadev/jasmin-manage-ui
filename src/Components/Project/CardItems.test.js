// src/Components/Project/CardItems.test.js

import React from "react";
import { render, screen, within } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import moment from "moment";
import {
  ProjectStatusListItem,
  ProjectConsortiumListItem,
  ProjectCollaboratorsListItem,
  ProjectCreatedAtListItem,
  ProjectTagItem,
  TagConsortiumItem,
} from "./CardItems";
import { useConsortia, useCurrentUser } from "../../api";
import { useNestedResource } from "../../rest-resource";

// Mock necessary hooks and modules
jest.mock("../../api", () => ({
  ...jest.requireActual("../../api"),
  useCurrentUser: jest.fn(),
  useConsortia: jest.fn(),
}));

jest.mock("../../rest-resource", () => ({
  ...jest.requireActual("../../rest-resource"),
  useNestedResource: jest.fn(),
}));

jest.mock("moment", () => () => ({ fromNow: () => "2 days ago" }));

describe("CardItems Components", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  // Tests for ProjectStatusListItem
  describe("ProjectStatusListItem", () => {
    // Helper function to render ProjectStatusListItem
    const renderProjectStatus = (status) => {
      const project = { data: { status } };
      render(<ProjectStatusListItem project={project} />);
    };

    it("Should render with status EDITABLE", () => {
      // Render the component with status 'EDITABLE'
      renderProjectStatus("EDITABLE");

      const listItem = screen.getByText(/Project is/);
      expect(listItem).toHaveClass(
        "list-group-item",
        "list-group-item-warning"
      );
      expect(listItem).toHaveTextContent(/EDITABLE/);
    });

    it("Should render with status UNDER_REVIEW", () => {
      // Render the component with status 'UNDER_REVIEW'
      renderProjectStatus("UNDER_REVIEW");

      const listItem = screen.getByText(/Project is/);
      expect(listItem).toHaveClass("list-group-item", "list-group-item-info");
      expect(listItem).toHaveTextContent(/UNDER_REVIEW/);
    });

    it("Should render with status COMPLETED", () => {
      // Render the component with status 'COMPLETED'
      renderProjectStatus("COMPLETED");

      const listItem = screen.getByText(/Project is/);
      expect(listItem).toHaveClass("list-group-item", "list-group-item-light");
      expect(listItem).toHaveTextContent(/COMPLETED/);
    });
  });

  // Tests for ProjectConsortiumListItem
  describe("ProjectConsortiumListItem", () => {
    // Helper function to render ProjectConsortiumListItem
    const renderProjectConsortium = (
      projectData,
      consortiaData,
      currentUserId
    ) => {
      useConsortia.mockReturnValue({
        fetchable: true,
        data: consortiaData,
        fetching: false,
        initialised: true,
      });
      useCurrentUser.mockReturnValue({
        data: { id: currentUserId },
        fetching: false,
        initialised: true,
      });
      render(
        <Router>
          <ProjectConsortiumListItem project={{ data: projectData }} />
        </Router>
      );
    };

    it("Should render correctly when user is consortium manager", () => {
      // Mock data where the current user is the consortium manager
      const projectData = { consortium: 1 };
      const consortiaData = {
        1: { data: { id: 1, name: "Consortium A", manager: { id: 2 } } },
      };

      renderProjectConsortium(projectData, consortiaData, 2);

      // Verify consortium name with link and manager message
      expect(screen.getByText(/Project belongs to/)).toBeInTheDocument();
      const consortiumLink = screen.getByRole("link", { name: "Consortium A" });
      expect(consortiumLink).toBeInTheDocument();
      expect(consortiumLink).toHaveAttribute("href", "/consortia/1");
      expect(
        screen.getByText("You are the consortium manager.")
      ).toBeInTheDocument();
    });

    it("Should render correctly when user is not consortium manager", () => {
      // Mock data where the current user is not the consortium manager
      const projectData = { consortium: 1 };
      const consortiaData = {
        1: { data: { id: 1, name: "Consortium A", manager: { id: 3 } } },
      };

      renderProjectConsortium(projectData, consortiaData, 2);

      // Verify consortium name without link and absence of manager message
      expect(screen.getByText(/Project belongs to/)).toBeInTheDocument();
      expect(
        screen.queryByRole("link", { name: "Consortium A" })
      ).not.toBeInTheDocument();
      expect(screen.getByText("Consortium A")).toBeInTheDocument();
      expect(
        screen.queryByText("You are the consortium manager.")
      ).not.toBeInTheDocument();
    });
  });

  // Tests for ProjectCollaboratorsListItem
  describe("ProjectCollaboratorsListItem", () => {
    // Helper function to render ProjectCollaboratorsListItem
    const renderProjectCollaborators = (
      projectData,
      collaboratorsData,
      currentUserId,
      initialised
    ) => {
      useCurrentUser.mockReturnValue({
        data: { id: currentUserId },
        fetching: false,
        initialised: true,
      });
      useNestedResource.mockReturnValue({
        data: collaboratorsData,
        initialised,
        fetching: false,
      });

      render(<ProjectCollaboratorsListItem project={{ data: projectData }} />);
    };

    it("Should handlde singular collaborator text correctly", () => {
      // Test case for exactly 1 collaborator
      const projectData = {
        num_collaborators: 1,
        current_user_role: "Member",
      };
      const collaboratorsData = {
        1: { data: { user: { id: 1 }, role: "Member" } },
      };

      renderProjectCollaborators(projectData, collaboratorsData, 1, true);

      // Verify singular form is used
      expect(screen.getByText(/Project has/)).toBeInTheDocument();
      expect(screen.getByText(/1 collaborator/)).toBeInTheDocument();
      expect(screen.queryByText(/collaborators/)).not.toBeInTheDocument();
    });

    it("Should handle vowel-starting roles correctly", () => {
      // Test case for role starting with vowel (e.g., "OWNER")
      const projectData = {
        num_collaborators: 1,
        current_user_role: "OWNER",
      };
      const collaboratorsData = {};

      renderProjectCollaborators(projectData, collaboratorsData, 1, false);

      // Verify "an" is used instead of "a"
      expect(screen.getByText(/You are an/)).toBeInTheDocument();
      expect(screen.getByText(/owner/)).toBeInTheDocument();
    });

    it("Should default to 0 collaborators when no data is available", () => {
      // Mock data with undefined num_collaborators and empty collaborators data
      const projectData = {};
      const collaboratorsData = {};

      renderProjectCollaborators(projectData, collaboratorsData, 1, true);

      // Verify that "0 collaborators" is displayed
      expect(screen.getByText(/Project has/)).toBeInTheDocument();
      expect(screen.getByText(/0 collaborators/)).toBeInTheDocument();
    });
  });

  // Tests for ProjectCreatedAtListItem
  describe("ProjectCreatedAtListItem", () => {
    it("Should render the creation date correctly", () => {
      const project = { data: { created_at: "2021-01-01T00:00:00Z" } };

      render(<ProjectCreatedAtListItem project={project} />);

      expect(screen.getByText(/Project was created/)).toBeInTheDocument();
      expect(screen.getByText("2 days ago")).toBeInTheDocument();
    });
  });

  // Tests for ProjectTagItem
  describe("ProjectTagItem", () => {
    it("Should render tags when they are available", () => {
      const tags = {
        data: {
          1: { data: { name: "Tag1" } },
          2: { data: { name: "Tag2" } },
        },
      };

      render(<ProjectTagItem tags={tags} />);

      expect(screen.getByText(/Tags:/)).toBeInTheDocument();
      expect(screen.getByText(/Tag1/)).toBeInTheDocument();
      expect(screen.getByText(/Tag2/)).toBeInTheDocument();
    });

    it("Should not render when there are no tags", () => {
      const tags = { data: {} };

      render(<ProjectTagItem tags={tags} />);

      expect(screen.queryByText(/Tags:/)).not.toBeInTheDocument();
    });
  });

  // Tests for TagConsortiumItem
  describe("TagConsortiumItem", () => {
    it("Should render consortium name when data is available", () => {
      const project = { data: { consortium: 1 } };
      const consortiaData = {
        1: { data: { name: "Consortium A" } },
      };

      useConsortia.mockReturnValue({
        fetchable: true,
        data: consortiaData,
        fetching: false,
        initialised: true,
      });

      render(<TagConsortiumItem project={project} />);

      expect(screen.getByText("Consortium A")).toBeInTheDocument();
    });

    it("Should show loading state when fetching consortia", () => {
      const project = { data: { consortium: 1 } };

      useConsortia.mockReturnValue({
        fetchable: true,
        fetching: true,
        initialised: false,
      });

      render(<TagConsortiumItem project={project} />);

      expect(screen.getByText("Loading consortia...")).toBeInTheDocument();
    });
  });

  // Snapshot Test
  describe("Snapshot tests", () => {
    it("Should match the snapshot for default rendering", () => {
      const project = { data: { status: "EDITABLE", consortium: 1 } };
      const consortiaData = {
        1: { data: { id: 1, name: "Consortium A", manager: { id: 1 } } },
      };

      // Mock hooks with initial data
      useConsortia.mockReturnValue({
        fetchable: true,
        data: consortiaData,
        fetching: false,
        initialised: true,
      });

      useCurrentUser.mockReturnValue({
        data: { id: 1 },
        fetching: false,
        initialised: true,
      });

      useNestedResource.mockReturnValue({
        data: {},
        initialised: false,
        fetching: false,
      });

      // Create component tree
      const component = renderer.create(
        <Router>
          <div>
            <ProjectStatusListItem project={project} />
            <ProjectConsortiumListItem project={project} />
            <ProjectCollaboratorsListItem project={project} />
            <ProjectCreatedAtListItem project={project} />
            <ProjectTagItem tags={{ data: {} }} />
          </div>
        </Router>
      );

      // Generate and verify snapshot
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
