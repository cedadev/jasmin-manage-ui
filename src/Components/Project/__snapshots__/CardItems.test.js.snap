// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItems Components Snapshot tests Should match the snapshot for default rendering 1`] = `
<div>
  <div
    className="list-group-item list-group-item-warning"
    disabled={false}
    onClick={[Function]}
  >
    Project is 
    <strong>
      EDITABLE
    </strong>
    .
  </div>
  <div
    className="list-group-item"
    disabled={false}
    onClick={[Function]}
  >
    <p
      className="mb-0"
    >
      Project belongs to
       
      <a
        className="font-weight-bold"
        href="/consortia/1"
        onClick={[Function]}
      >
        Consortium A
      </a>
      .
    </p>
    <p
      className="mt-2 mb-0"
    >
      <strong>
        You are the consortium manager.
      </strong>
    </p>
  </div>
  <div
    className="list-group-item"
    disabled={false}
    onClick={[Function]}
  >
    <p
      className="mb-0"
    >
      Project has
       
      <strong
        project={
          {
            "data": {
              "consortium": 1,
              "status": "EDITABLE",
            },
          }
        }
      >
        0
         collaborator
        s
      </strong>
      .
    </p>
  </div>
  <div
    className="list-group-item"
    disabled={false}
    onClick={[Function]}
  >
    Project was created 
    <strong>
      2 days ago
    </strong>
    .
  </div>
  
</div>
`;
